### PR TITLE
tweak Programs locking + minor refactoring

### PR DESCRIPTION
**Pull request by pattyshack**
- reduce Get's critical section (don't hold lock when looking up parent's entry)
- guard HasChanges

cc: https://github.com/dapperlabs/flow-internal/issues/1722

### CODE REVIEW

**Review by SaveTheRbtz — APPROVED** [review]

**Comment by turbolent** [comment]
@pattyshack by creating PRs from branches in this repo, CI can run benchmarks

**Comment by pattyshack** [comment]
@turbolent Is the standard procedure is to create branches directly in onflow/flow-go ?  I'm happy to comply if that's the standard, but this practice seems a bit surprising since it excludes external developers from contributing to the repo (or at least creates a barrier of entry).

Perhaps the right thing to do is to fix CI to run benchmarks regardless of which repo the PR originated from?

**Comment by pattyshack** [comment]
@Kay-Zee looks like the ci benchstat lacks "pull-request"-write permission from forks due to GITHUB_TOKEN settings.  would it be possible to enable that by default (or is that too much of a security concern)?

File: fvm/programs/programs.go
Comment (turbolent) [inline fvm/programs/programs.go:43]: Was there a particular reason this was a function in the first place?

File: fvm/programs/programs.go
Comment (turbolent) [inline fvm/programs/programs.go:96]: Good idea to refactor this common code out, and nice job naming it well 👍

**Review by turbolent — APPROVED** [review]
LGTM! Maybe wait for a review from e.g. @ramtinms (@m4ksio is OOO)

**Review by janezpodhostnik — APPROVED** [review]
LGTM as well. Nice change!

**Comment by pattyshack** [comment]
bors retry

File: fvm/programs/programs.go
Comment (ramtinms) [reply fvm/programs/programs.go:43]: I vaguely remember that there was a reason for this to be a func, but not in details, I'm okey we change it for now and when @m4ksio is back we would ask him.

**Review by ramtinms — APPROVED** [review]
Looks good to me

File: fvm/programs/programs.go
Comment (m4ksio) [reply fvm/programs/programs.go:43]: I think we wanted to prevent long chains of programs. Having a function gave us ability to copy entries from parents up, but I don't think we ever used it

File: fvm/programs/programs.go
Comment (m4ksio) [inline fvm/programs/programs.go:80]: Was there a reason to move to a pointer? Isn't this increasing allocations count, instead of keeping data inside a map?

**Pull request merged**
